### ISSUE TITLE
Cron task cleanupDatabase() causes warning if DB table not exist

### ISSUE THREAD

**Issue by imsadhappy** [arnowelzel/lightbox-photoswipe#18]
Hi.

Exception message:
`WordPress-Datenbank-Fehler Table 'colcms_tickets.tkt_lightbox_photoswipe_img' doesn't exist für Abfrage DELETE FROM tkt_lightbox_photoswipe_img where created<("2020-04-12 15:39:32") von do_action_ref_array('lbwps_cleanup'), WP_Hook->do_action, WP_Hook->apply_filters, LightboxPhotoSwipe->cleanupDatabase`

I've located the place at lightbox-photoswipe.php:1034

```
    function cleanupDatabase()
    {
        global $wpdb;

        $table_img = $wpdb->prefix . 'lightbox_photoswipe_img';
        $date = strftime('%Y-%m-%d %H:%M:%S', time()-86400);
        $sql = "DELETE FROM $table_img where created<(\"$date\")";
        $wpdb->query($sql);
    } 
```
I would suggest checking if table exists before executing db query like so:

```
    function cleanupDatabase()
    {
        global $wpdb;

        $table_img = $wpdb->prefix . 'lightbox_photoswipe_img';

        if ($wpdb->get_var( $wpdb->prepare( "SHOW TABLES LIKE %s", $table_img ) ) === $table_img) {
            $date = strftime('%Y-%m-%d %H:%M:%S', time()-86400);
            $sql = "DELETE FROM $table_img where created<(\"$date\")";
            $wpdb->query($sql);
        }
    }  
```

**Comment by arnowelzel** [comment]
Thanks for pointing this out. Why `prepare()` and not `query()`?

However the error is not expected, since the table should exist, otherwise other errors will occur as well. `cleanupDatebase()` is not the only place which uses the database. `outputCallbackProperties()` needs the database as well. Also the table is created before any cron job runs.

So the question is: why does the table not exist on your site? Did this error occur only once during plugin installation? Then it might be a side effect, that activating the cron job happens before `init()` got triggered.

**Comment by arnowelzel** [comment]
Fixed with 3e8dd9148d6fabb3a43fdf42c2c43d61caa881bd.

**Comment by arnowelzel** [comment]
After some more investigation it turned out, that release 2.80 had more severe issues.

This got all fixed with https://github.com/arnowelzel/lightbox-photoswipe/commit/25e2647ff974d601fdf5159a874b1c9df8b1a977

**Comment by imsadhappy** [comment]
Hi. This issue emmerged again in the version 2.90 with a new twist.

`Notice: Undefined variable: table_img in /home/colcms/developtimization.com/tickets/wp-content/plugins/lightbox-photoswipe/lightbox-photoswipe.php on line 1074`

The code fragment now looks like so:

```
    function cleanupDatabase()
    {
        global $wpdb;

        $date = strftime( '%Y-%m-%d %H:%M:%S', time() - 86400 );
        $sql  = "DELETE FROM $table_img where created<(\"$date\")";
        $wpdb->query( $sql );
    }
```

It obviously lacks declaration of `$table_img`. Also, I would definetely recommend checking that table exists (from the conversation above) - it won't hurt.

All together, should it look like so now (?):

```
    function cleanupDatabase()
    {
        global $wpdb;

        $table_img = $wpdb->prefix . 'lightbox_photoswipe_img';

        if ($wpdb->get_var( $wpdb->prepare( "SHOW TABLES LIKE %s", $table_img ) ) === $table_img) {
            $date = strftime('%Y-%m-%d %H:%M:%S', time()-86400);
            $sql = "DELETE FROM $table_img where created<(\"$date\")";
            $wpdb->query($sql);
        }
    } 
```

**Comment by arnowelzel** [comment]
Maybe there is a misunderstanding: The table **must** exist. Otherwise there are **way** more serious problems since the same table is also used in https://github.com/arnowelzel/lightbox-photoswipe/blob/2.90/lightbox-photoswipe.php#L557 and https://github.com/arnowelzel/lightbox-photoswipe/blob/2.90/lightbox-photoswipe.php#L529.

Also the actual code of 2.90 is already  different - see here:
https://github.com/arnowelzel/lightbox-photoswipe/blob/2.90/lightbox-photoswipe.php#L1069

**Comment by imsadhappy** [comment]
The issue is two-fold. Please have a closer look at current function. First (important): $table_img is not declared - the code will do nothing. This one has nothing to do with whether the table exists or not.

Second (not so important): about checking if table exists from the initial comment. The fact that it **must** exist, doesn't mean that it **will** exist. There might be some problem while creating it, it can be dropped from aside (by client, or by some other plugin).

**Comment by arnowelzel** [comment]
Frist: please have a close lock a the code - I already mentioned that - there is no `$table_img` any longer in version 2.90!

https://github.com/arnowelzel/lightbox-photoswipe/blob/2.90/lightbox-photoswipe.php#L1069

```
    function cleanupDatabase()
    {
        global $wpdb;

        $date = strftime( '%Y-%m-%d %H:%M:%S', time() - 86400 );
        $sql  = "DELETE FROM $table_img where created<(\"$date\")";
        $wpdb->query( $sql );
    }            
```

So please read my comment again and with ALL references I mentioned - thank you!

**Comment by arnowelzel** [comment]
And also check my commit https://github.com/arnowelzel/lightbox-photoswipe/commit/25e2647ff974d601fdf5159a874b1c9df8b1a977 before commenting again - thank you!

**Comment by imsadhappy** [comment]
You did not define variable. It's not about table exists or not. Your sql will execute like "DELETE FROM undefined where created ..."; $table_img or whatever call it - you missed `$table_img = $wpdb->prefix . 'lightbox_photoswipe_img';` declaration.

**Comment by arnowelzel** [comment]
Sorry - but it seems you don't understand it.

The original issue you reported was caused by an error in the table creation. So the table got not created and **that** was the bug. The bug was **not** to miss a check in `cleanupDatabase()`.

Please install the current version from https://wordpress.org/plugins/lightbox-photoswipe/ and test it. Anything else won't help here. Thank you!

**Comment by arnowelzel** [comment]
Last error fixed with 69a8b4da45a518381f85c73ab124845abf33e6ef. In fact the code is now reverted as it was originally.
